plugins/kotlin/j2k/shared/src/org/jetbrains/kotlin/nj2k/PostprocessorExtensionsRunner.kt

```diff
@@ -0,0 +1,45 @@
+// Copyright 2000-2024 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
+package org.jetbrains.kotlin.nj2k
+
+import com.intellij.openapi.progress.*
+import com.intellij.openapi.project.Project
+import org.jetbrains.kotlin.idea.codeinsight.utils.commitAndUnblockDocument
+import org.jetbrains.kotlin.j2k.J2kPostprocessorExtension
+import org.jetbrains.kotlin.psi.KtFile
+
+/**
+ * Before conversion, runs all registered custom postprocessor extensions (i.e. classes implementing `J2kPostprocessorExtension` and
+ * registered in their parent plugin's xml file).
+ */
+object PostprocessorExtensionsRunner {
+    private const val PHASE_NAME = "Custom Postprocessing"
+
+    fun runRegisteredPostprocessors(project: Project, ktFiles: List<KtFile>, postprocessorExtensions: List<J2kPostprocessorExtension>) {
+        if (postprocessorExtensions.isEmpty()) return
+        val postprocessorsCount = postprocessorExtensions.size
+        ProgressManager.progress(PHASE_NAME, "Found $postprocessorsCount postprocessors to run on Java files before conversion")
+        for ((i, postprocessor) in postprocessorExtensions.withIndex()) {
+            ProgressManager.checkCanceled()
+            ProgressManager.progress(PHASE_NAME, "Running postprocessor ${i + 1}/$postprocessorsCount")
+            try {
+                runBlockingCancellable {
```

Comment:
@darthorimar this runs just fine in the test IDE, but when I run my postprocessor unit test `testClassWithParamsAndProperties`, I get different types of undesirable behavior, depending on how I launch the coroutine:
1. `runBlocking` --> runs forever
2. `runBlocking(newSingleThreadContext("MyThread"))` --> runs forever
3. `runBlockingCancellable` --> "ERROR: There is no ProgressIndicator or Job in this thread, the current job is not cancellable."
4. `runBlockingMaybeCancellable` --> "java.lang.IllegalStateException: Switching to Dispatchers.EDT from `runBlockingCancellable`  inside in a read-action leads to possible deadlock."

This seems especially weird since the test case is just running on the main thread.